ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -683,7 +717,177 @@ public void handleReplicaMetadataRequest(NetworkRequest request) throws IOExcept
+    metrics.replicateBlobProcessingTimeInMs.update(processingTime);
+    requestResponseChannel.sendResponse(response, request,
+        new ServerNetworkResponseMetrics(metrics.replicateBlobResponseQueueTimeInMs, metrics.replicateBlobSendTimeInMs,
+            metrics.replicateBlobTotalTimeInMs, null, null, totalTimeSpent));
+  }
+
+  /**
+   * to get the Blob from the remote replica
+   * @param replicateBlobRequest the {@link ReplicateBlobRequest}
+   * @param blobId blobId of the ReplicateBlob request
+   * @param remoteHostName the name of the remote host to get the blob from
+   * @param remoteHostPort the port of the remote host to get the blob from
+   * @return a pair of the {@link ServerErrorCode} and the {@link GetResponse}.
+   */
+  private Pair<ServerErrorCode, GetResponse> getBlobFromRemoteReplicate(ReplicateBlobRequest replicateBlobRequest,
+      BlobId blobId, String remoteHostName, int remoteHostPort) {
+    DataNodeId remoteDataNode = clusterMap.getDataNodeId(remoteHostName, remoteHostPort);
+    if (remoteDataNode == null) {
+      logger.error("ReplicateBlobRequest {} couldn't find the remote host {} {} in the clustermap.", blobId,
+          remoteHostName, remoteHostPort);
+      return new Pair(ServerErrorCode.Replica_Unavailable, null);
+    }
+
+    // ON_DEMAND_REPLICATION_TODO: Add configuration as replicationConfig.replicationConnectionPoolCheckoutTimeoutMs
+    final int connectionPoolCheckoutTimeoutMs = 1000;
+    List<PartitionRequestInfo> partitionRequestInfoList = new ArrayList<>();
+    List<BlobId> blobIds = new ArrayList<>();
+    blobIds.add(blobId);
+    PartitionRequestInfo partitionInfo = new PartitionRequestInfo(blobId.getPartition(), blobIds);
+    partitionRequestInfoList.add(partitionInfo);
+    final String clientId = ON_DEMAND_REPLICATION_CLIENTID_PREFIX + currentNode.getHostname();
+
+    try {
+      // Get the Blob including expired and deleted.
+      GetRequest getRequest = new GetRequest(replicateBlobRequest.getCorrelationId(), clientId, MessageFormatFlags.All,
+          partitionRequestInfoList, GetOption.Include_All);
+
+      // get the blob from the remote replica
+      ConnectedChannel connectedChannel =
+          connectionPool.checkOutConnection(remoteHostName, remoteDataNode.getPortToConnectTo(),
+              connectionPoolCheckoutTimeoutMs);
+      ChannelOutput channelOutput = connectedChannel.sendAndReceive(getRequest);
+      GetResponse getResponse = GetResponse.readFrom(channelOutput.getInputStream(), clusterMap);
+      if (getResponse.getError() != ServerErrorCode.No_Error) {
+        logger.error("ReplicateBlobRequest failed to get blob {} from the remote node {} {} {}", blobId, remoteHostName,
+            remoteHostPort, getResponse.getError());
+        return new Pair(getResponse.getError(), getResponse);
+      }
+      if ((getResponse.getPartitionResponseInfoList() == null) || (getResponse.getPartitionResponseInfoList().size()
```

Comment:
nit: we don't have to add parenthese around these two statements.

Reply:
It's my coding style. It's for easy reading and "no need to worry about precedence". 
I hope it's ok that I keep it.

Reply:
Please condense verbose booleans into helpers like isSomethingXX

Reply:
Do you mean wrapping the "getResponse.getPartitionResponseInfoList() == null) || (getResponse.getPartitionResponseInfoList().size()==0"?
I feel it's a little bit overkilled.

Reply:
Yes, it would be readable to have such a helper even if its a one-liner. I can never understand long booleans and they keep growing.

Reply:
To my opinion, two condition is not long. And it is not used much in other places, so I skip this change.